File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/settings/ShowsDistillationSettings.java
/*
 * Copyright 2013 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.battlelancer.seriesguide.settings;

import android.content.Context;
import android.preference.PreferenceManager;

import com.battlelancer.seriesguide.Constants;

import static com.battlelancer.seriesguide.provider.SeriesContract.Shows;

/**
 * Provides settings used to filter and sort displayed shows in
 * {@link com.battlelancer.seriesguide.ui.ShowsFragment}.
 */
public class ShowsDistillationSettings {

    public static String KEY_SORT_ORDER = "com.battlelancer.seriesguide.sort.order";
    public static String KEY_SORT_FAVORITES_FIRST = "com.battlelancer.seriesguide.sort.favoritesfirst";
    public static String KEY_FILTER_FAVORITES = "com.battlelancer.seriesguide.filter.favorites";
    public static String KEY_FILTER_UNWATCHED = "com.battlelancer.seriesguide.filter.unwatched";
    public static String KEY_FILTER_UPCOMING = "com.battlelancer.seriesguide.filter.upcoming";
    public static String KEY_FILTER_HIDDEN = "com.battlelancer.seriesguide.filter.hidden";

    /**
     * Builds an appropriate SQL sort statement for sorting shows.
     */
    public static String getSortQuery(int sortOrderId, boolean isSortFavoritesFirst) {
        StringBuilder query = new StringBuilder();

        if (isSortFavoritesFirst) {
            query.append(ShowsSortOrder.FAVORITES_FIRST_PREFIX);
        }

        if (sortOrderId == ShowsSortOrder.TITLE_REVERSE_ID) {
            query.append(ShowsSortOrder.TITLE_REVERSE);
        } else if (sortOrderId == ShowsSortOrder.EPISODE_ID) {
            query.append(ShowsSortOrder.EPISODE);
        } else if (sortOrderId == ShowsSortOrder.EPISODE_REVERSE_ID) {
            query.append(ShowsSortOrder.EPISODE_REVERSE);
        } else {
            query.append(ShowsSortOrder.TITLE);
        }

        return query.toString();
    }

    /**
     * Returns the id as of
     * {@link com.battlelancer.seriesguide.settings.ShowsDistillationSettings.ShowsSortOrder}
     * of the current show sort order.
     */
    public static int getSortOrderId(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context).getInt(KEY_SORT_ORDER, 0);
    }

    public static boolean isSortFavoritesFirst(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(KEY_SORT_FAVORITES_FIRST, true);
    }

    public static boolean isFilteringFavorites(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(KEY_FILTER_FAVORITES, false);
    }

    public static boolean isFilteringUnwatched(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(KEY_FILTER_UNWATCHED, false);
    }

    public static boolean isFilteringUpcoming(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(KEY_FILTER_UPCOMING, false);
    }

    public static boolean isFilteringHidden(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getBoolean(KEY_FILTER_HIDDEN, false);
    }

    /**
     * Used by {@link com.battlelancer.seriesguide.ui.ShowsFragment} loader to sort the list of
     * shows.
     */
    public interface ShowsSortOrder {
        // alphabetical by title
        String TITLE = Shows.TITLE + " COLLATE NOCASE ASC";
        // reverse alphabetical by title
        String TITLE_REVERSE = Shows.TITLE + " COLLATE NOCASE DESC";
        // by next episode air time, oldest first
        String EPISODE = Shows.NEXTAIRDATEMS + " ASC," + Shows.AIRSTIME + " ASC,"
                + Shows.TITLE + " COLLATE NOCASE ASC";
        // by next episode air time, newest first
        String EPISODE_REVERSE = Shows.NEXTAIRDATEMS + " DESC," + Shows.AIRSTIME + " ASC,"
                + Shows.TITLE + " COLLATE NOCASE ASC";
        // add as prefix to sort favorites first
        String FAVORITES_FIRST_PREFIX = Shows.FAVORITE + " DESC,";
        // ids used for storing in preferences
        int TITLE_ID = 0;
        int TITLE_REVERSE_ID = 1;
        int EPISODE_ID = 2;
        int EPISODE_REVERSE_ID = 3;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/ShowsDistillationSettings.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2019-2025 Uwe Trottmann

package com.battlelancer.seriesguide.shows

import android.content.Context
import androidx.core.content.edit
import androidx.preference.PreferenceManager
import com.battlelancer.seriesguide.appwidget.ListWidgetProvider
import com.battlelancer.seriesguide.provider.SeriesGuideContract.SgShow2Columns
import com.battlelancer.seriesguide.settings.DisplaySettings
import kotlinx.coroutines.flow.MutableStateFlow

/**
 * Provides settings used to filter and sort displayed shows in [ShowsFragment].
 */
class ShowsDistillationSettings(
    private val context: Context
) {

    /**
     * Initially the current value, emits when the filter settings were changed with [saveFilters].
     */
    val showFilters = MutableStateFlow(ShowFilters.fromSettings(context))

    /**
     * Initially the current value, emits when the sort order was changed with [saveSortOrder].
     */
    val sortOrder = MutableStateFlow(ShowSortOrder.fromSettings(context))

    fun saveFilters(showFilters: ShowFilters) {
        // Save setting
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putInt(KEY_FILTER_FAVORITES, showFilters.isFilterFavorites.mapFilterState())
            putInt(KEY_FILTER_UNWATCHED, showFilters.isFilterUnwatched.mapFilterState())
            putInt(KEY_FILTER_UPCOMING, showFilters.isFilterUpcoming.mapFilterState())
            putInt(KEY_FILTER_HIDDEN, showFilters.isFilterHidden.mapFilterState())
            putInt(KEY_FILTER_CONTINUING, showFilters.isFilterContinuing.mapFilterState())
        }
        // Broadcast new value
        this.showFilters.value = showFilters
    }

    fun saveSortOrder(showSortOrder: ShowSortOrder) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putInt(KEY_SORT_ORDER, showSortOrder.sortOrderId)
            putBoolean(
                KEY_SORT_FAVORITES_FIRST,
                showSortOrder.isSortFavoritesFirst
            )
            putBoolean(
                DisplaySettings.KEY_SORT_IGNORE_ARTICLE,
                showSortOrder.isSortIgnoreArticles
            )
        }

        // broadcast new sort order
        sortOrder.value = showSortOrder

        // Note: List widgets continue to share the ignore articles setting
        if (showSortOrder.changedIgnoreArticles) {
            // refresh all list widgets
            ListWidgetProvider.notifyDataChanged(context)
        }
    }

    data class ShowFilters(
        val isFilterFavorites: Boolean?,
        val isFilterUnwatched: Boolean?,
        val isFilterUpcoming: Boolean?,
        val isFilterHidden: Boolean?,
        val isFilterContinuing: Boolean?
    ) {
        fun isAnyFilterEnabled(): Boolean {
            return isFilterFavorites != null || isFilterUnwatched != null
                    || isFilterUpcoming != null || isFilterHidden != null
                    || isFilterContinuing != null
        }

        companion object {
            fun default(): ShowFilters {
                // Exclude hidden, all others disabled.
                return ShowFilters(null, null, null, false, null)
            }

            fun fromSettings(context: Context): ShowFilters {
                val prefs = PreferenceManager.getDefaultSharedPreferences(context)
                return ShowFilters(
                    prefs.getInt(KEY_FILTER_FAVORITES, FILTER_DISABLED).mapFilterState(),
                    prefs.getInt(KEY_FILTER_UNWATCHED, FILTER_DISABLED).mapFilterState(),
                    prefs.getInt(KEY_FILTER_UPCOMING, FILTER_DISABLED).mapFilterState(),
                    // exclude hidden shows by default
                    prefs.getInt(KEY_FILTER_HIDDEN, FILTER_EXCLUDE).mapFilterState(),
                    prefs.getInt(KEY_FILTER_CONTINUING, FILTER_DISABLED).mapFilterState()
                )
            }
        }
    }

    data class ShowSortOrder(
        val sortOrderId: Int,
        val isSortFavoritesFirst: Boolean,
        val isSortIgnoreArticles: Boolean,
        val changedIgnoreArticles: Boolean
    ) {
        companion object {

            const val TITLE_ID = 0
            // @deprecated Only supporting alphabetical sort order going forward.
            // int TITLE_REVERSE_ID = 1;
            const val OLDEST_EPISODE_ID = 2
            const val LATEST_EPISODE_ID = 3
            const val LAST_WATCHED_ID = 4
            const val LEAST_REMAINING_EPISODES_ID = 5
            const val STATUS = 6

            fun fromSettings(context: Context): ShowSortOrder {
                val prefs = PreferenceManager.getDefaultSharedPreferences(context)
                return ShowSortOrder(
                    prefs.getInt(KEY_SORT_ORDER, TITLE_ID),
                    prefs.getBoolean(KEY_SORT_FAVORITES_FIRST, true),
                    DisplaySettings.isSortOrderIgnoringArticles(context),
                    false
                )
            }
        }
    }

    companion object {

        private const val KEY_SORT_ORDER = "com.battlelancer.seriesguide.sort.order"
        private const val KEY_SORT_FAVORITES_FIRST =
            "com.battlelancer.seriesguide.sort.favoritesfirst"
        private const val KEY_FILTER_FAVORITES = "seriesguide.show_filter.favorites"
        private const val KEY_FILTER_UNWATCHED = "seriesguide.show_filter.unwatched"
        private const val KEY_FILTER_UPCOMING = "seriesguide.show_filter.upcoming"
        private const val KEY_FILTER_HIDDEN = "seriesguide.show_filter.hidden"
        private const val KEY_FILTER_CONTINUING = "seriesguide.show_filter.continuing"

        private const val FILTER_INCLUDE = 1
        private const val FILTER_EXCLUDE = -1
        private const val FILTER_DISABLED = 0

        private fun Int.mapFilterState(): Boolean? {
            return when (this) {
                FILTER_INCLUDE -> true
                FILTER_EXCLUDE -> false
                else -> null
            }
        }

        private fun Boolean?.mapFilterState(): Int {
            return when (this) {
                null -> FILTER_DISABLED
                true -> FILTER_INCLUDE
                false -> FILTER_EXCLUDE
            }
        }

        private object SgShow2SortQuery {
            // by oldest next episode, then continued first (for no next episode)
            const val OLDEST_EPISODE =
                "${SgShow2Columns.NEXTAIRDATEMS} ASC,${SgShow2Columns.SORT_STATUS},"

            // by latest next episode, then continued first (for no next episode)
            const val LATEST_EPISODE = "${SgShow2Columns.SORT_LATEST_EPISODE_THEN_STATUS},"

            // by latest watched first
            const val LAST_WATCHED = "${SgShow2Columns.LASTWATCHED_MS} DESC,"

            // by least episodes remaining to watch, then continued first (for no remaining episode)
            const val REMAINING_EPISODES =
                "${SgShow2Columns.UNWATCHED_COUNT} ASC,${SgShow2Columns.SORT_STATUS},"

            const val STATUS = "${SgShow2Columns.STATUS} DESC,"

            // add as prefix to sort favorites first
            const val FAVORITES_FIRST = "${SgShow2Columns.FAVORITE} DESC,"
        }

        /**
         * Builds an appropriate SQL sort statement for sorting SgShow2 table results.
         */
        @JvmStatic
        fun getSortQuery2(
            sortOrderId: Int,
            isSortFavoritesFirst: Boolean,
            isSortIgnoreArticles: Boolean
        ): String {
            val query = StringBuilder()

            if (isSortFavoritesFirst) {
                query.append(SgShow2SortQuery.FAVORITES_FIRST)
            }

            when (sortOrderId) {
                ShowSortOrder.OLDEST_EPISODE_ID -> query.append(SgShow2SortQuery.OLDEST_EPISODE)
                ShowSortOrder.LATEST_EPISODE_ID -> query.append(SgShow2SortQuery.LATEST_EPISODE)
                ShowSortOrder.LAST_WATCHED_ID -> query.append(SgShow2SortQuery.LAST_WATCHED)
                ShowSortOrder.LEAST_REMAINING_EPISODES_ID -> query.append(SgShow2SortQuery.REMAINING_EPISODES)
                ShowSortOrder.STATUS -> query.append(SgShow2SortQuery.STATUS)
            }
            // always sort by title at last
            query.append(
                if (isSortIgnoreArticles) {
                    SgShow2Columns.SORT_TITLE_NOARTICLE
                } else {
                    SgShow2Columns.SORT_TITLE
                }
            )

            return query.toString()
        }
    }
}
